Remove fetch-client events

chestercharles opened the pull request:
Pretty much everything was just a matter of stripping out events.

The only thing I actually had to change was the oauth plugin implementation, since the refresh logic used events. It now just uses a promise.

chestercharles commented on src/plugins/oauth.js:
```diff
@@ -1,29 +1,19 @@
-    } else {
-      request.headers.set(
-        'Authorization',
-        `Bearer ${this.client.getBearerToken()}`
```
Code in this file was the only thing changed/added (as opposed to removed)

spruce-bruce commented on src/plugins/oauth.js:
```diff
@@ -1,29 +1,19 @@
-        'Authorization',
-        `Bearer ${this.client.getBearerToken()}`
+      const refreshed = await new Promise(
+        (resolve) => (this.client.refreshed = resolve)
```
I haven't run this to see, so correct me if I'm wrong:

If you have multiple requests coming in while an id token is being refreshed then `this.client.refreshed` will be overwritten for each one of those incoming requests. When the refresh is completed and `this.client.refreshed()` is called only the last request in the queue will be resolved, leaving the previous requests permanently unresolved.

IIRC we tried writing tests for cases like this, but it's difficult to test for

chestercharles replied:
Is the desired behavior that the subsequent requests wait until the refresh occurs, and then proceed? Instead of instantiating a new promise in each `onStart` we could check for any existing promise and await that.

spruce-bruce replied:
Yeah.

Basically the fetch-client is handling a stream of requests using the authorization token that it knows about. One of those requests will eventually return a 401 because the token has expired, and and the oauth plugin goes and gets a new token. While the oauth plugin is doing this work the stream of requests is still coming in, and they will all return a 401 if we don't pause them and wait for the refresh to complete.

FURTHERMORE, there might be multiple pending requests using an expired token. All of these will respond with a 401, but only the first one should trigger the token refresh. Any requests that return a 401 during the refresh should be retried.

We've been emulating a queue of requests using the event handler, essentially. During the refresh we don't execute requests right away, we add event handlers that execute the requests to the queue, and when the "refreshed" event fires all of those handlers are called. It's pretty convoluted frankly.

I think checking for any existing promise and awaiting that seems like it might work, but I'm worried that it'd be possible to introduce a race condition here. I think we tried something like that and it turned out to be hard. IIRC `this.client.refreshPromise` was a thing we set and tried awaiting

spruce-bruce commented:
The original issue: https://github.com/synapsestudios/fetch-client/issues/7

That sat open for a long time. Gordon had that cool observable example to solve this but we didn't want to get into the rxjs dependency at the time. Might be a good rewrite for this lib, but obviously not for this particular issue, and from that issue you can see the original solution zw/zc worked on getting merged.

chestercharles commented:
It's been awhile on this PR, but I've refactored the auth plugin so that it uses a promise instead of the client's event emitter. I believe that I've removed the problem where the refresh request will get overwritten now.

If we are starting a request, we check if the plugin is refreshing the token. If it is, then we wait until the refresh is complete. Otherwise, we continue.

On completion of a request, we check to see if the response is a 401 and whether or not we are currently refreshing the token.

Is it a 401 and we are NOT refreshing the token?
Initiate that refresh request and chain on the stuff we want to happen on successful refresh (add used refresh token, invoke onRefreshResponse). We set that whole promise to `this.client.refreshRequest`. Then wait for `this.client.refreshResponse` to resolve. If it was successful, then retry with a cloned request.

Is it a 401 and we ARE refreshing the token?
Then wait for `this.client.refreshResponse` to resolve. If it was successful, then retry with a cloned request.

Is it NOT a 401?
Then just return the response.

One thing I'm not sure about - if we start a request during refreshing and the refresh fails, what should happen? Currently, we return the request without setting any auth headers. The code in this PR will set the auth headers to a (presumably) expired auth token. Either way it seem like we would expect a 401 from the server.

spruce-bruce reviewed (commented):
Sweet!

There's a bunch of usage documentation that needs to be updated in the README. Is that on your radar?

chestercharles commented:
I updated the README to omit all the event's API stuff - is that sufficient?

spruce-bruce reviewed (approved):
Oh yeah, I missed that in my review earlier. LGTM!
